src/ott/neural/solvers/base_solver.py

```diff
@@ -0,0 +1,480 @@
+      )
+    else:
+      indices = jnp.arange(batch_size)
+    tmat_adapted = tmat[indices]
+    indices_per_row = jax.vmap(
+        lambda tmat_adapted: jax.random.choice(
+            key=key,
+            a=jnp.arange(batch_size),
+            p=tmat_adapted,
+            shape=(k_samples_per_x,)
+        ),
+        in_axes=0,
+        out_axes=0,
+    )(
+        tmat_adapted
+    )
+
+    indices_source = jnp.repeat(indices, k_samples_per_x)
+    indices_target = jnp.reshape(
+        indices_per_row % tmat.shape[1], (batch_size * k_samples_per_x,)
+    )
+    return tuple(
+        jnp.reshape(b[indices_source], (k_samples_per_x, batch_size,
+                                        -1)) if b is not None else None
+        for b in source_arrays
+    ), tuple(
+        jnp.reshape(b[indices_target, :], (k_samples_per_x, batch_size,
+                                           -1)) if b is not None else None
+        for b in target_arrays
+    )
+
+  def _get_sinkhorn_match_fn(
+      self,
+      ot_solver: Any,
+      epsilon: float = 1e-2,
+      cost_fn: costs.CostFn = costs.SqEuclidean(),
+      scale_cost: Union[bool, int, float,
+                        Literal["mean", "max_norm", "max_bound", "max_cost",
+                                "median"]] = "mean",
+      tau_a: float = 1.0,
+      tau_b: float = 1.0,
+      *,
+      filter_input: bool = False,
+  ) -> Callable:
+
+    def match_pairs(
+        x: jnp.ndarray, y: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+      geom = pointcloud.PointCloud(
+          x, y, epsilon=epsilon, scale_cost=scale_cost, cost_fn=cost_fn
+      )
+      return ot_solver(
+          linear_problem.LinearProblem(geom, tau_a=tau_a, tau_b=tau_b)
+      ).matrix
+
+    def match_pairs_filtered(
+        x_lin: jnp.ndarray, x_quad: jnp.ndarray, y_lin: jnp.ndarray,
+        y_quad: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+      geom = pointcloud.PointCloud(
+          x_lin, y_lin, epsilon=epsilon, scale_cost=scale_cost, cost_fn=cost_fn
+      )
+      return ot_solver(
+          linear_problem.LinearProblem(geom, tau_a=tau_a, tau_b=tau_b)
+      ).matrix
+
+    return match_pairs_filtered if filter_input else match_pairs
+
+  def _get_gromov_match_fn(
+      self,
+      ot_solver: Any,
+      cost_fn: Union[Any, Mapping[str, Any]],
+      scale_cost: Union[Union[bool, int, float,
+                              Literal["mean", "max_norm", "max_bound",
+                                      "max_cost", "median"]],
+                        Dict[str, Union[bool, int, float,
+                                        Literal["mean", "max_norm", "max_bound",
+                                                "max_cost", "median"]]]],
+      tau_a: float,
+      tau_b: float,
+      fused_penalty: float,
+  ) -> Callable:
+    if isinstance(cost_fn, Mapping):
+      assert "x_cost_fn" in cost_fn
```

Comment:
`cost_fn_x` or `cost_fn_xx` sounds more clear to me (agrees with math notation more as well, as well as `geom_xx`)